src/common/lazy_alloc_array.h

```diff
@@ -12,21 +12,23 @@
 #include <mutex>
 #include <array>
 #include <vector>
+#include <atomic>
 
 namespace mxnet {
 namespace common {
 
 template<typename TElem>
 class LazyAllocArray {
  public:
+  LazyAllocArray();
   /*!
    * \brief Get element of corresponding index,
    *  if it is not created create by creator
    * \param index the array index position
    * \param creator a lambda function to create new element when needed.
    */
   template<typename FCreate>
-  inline TElem* Get(int index, FCreate creator);
+  inline std::shared_ptr<TElem> Get(int index, FCreate creator);
```

Comment:
Because during shutdown, this pointer can be stale, since this will be called from a different thread than what's calling Clear() on NDLazyArray. This was experienced.